### ISSUE TITLE
Could not reach the service. Is the service running and connected to the same network as the proxy

### ISSUE THREAD

**Issue by getvivekv** [vfarcic/docker-flow-proxy#115]
Hi,

Testing with over 30+ virtualhost/containers on 5 nodes. Often I am getting this error message.

```
2017/01/20 12:54:55 Processing request /v1/docker-flow-proxy/reconfigure?distribute=true&port=80&serviceDomain=%2Ammane-sl.domain.com&serviceName=mmane-sl&servicePath=%2F

2017/01/20 12:54:55 Sending distribution request to http://10.0.1.78:8080/v1/docker-flow-proxy/reconfigure?distribute=false&port=80&serviceDomain=%2Ammane-sl.domain.com&serviceName=mmane-sl&servicePath=%2F

2017/01/20 12:54:55 Processing request /v1/docker-flow-proxy/reconfigure?distribute=false&port=80&serviceDomain=%2Ammane-sl.domain.com&serviceName=mmane-sl&servicePath=%2F

```

I am using a script to stand up docker services, and yes this service is connected to proxy network. 

```
[root@dev-swarm-1 ~]# docker network ls 
NETWORK ID          NAME                DRIVER              SCOPE
aqz5md847p95        appcenter           overlay             swarm               
c8a1cd3790bf        bridge              bridge              local               
b2c59f7f8272        docker_gwbridge     bridge              local               
dab396a5f991        host                host                local               
244ob6bfjdwh        ingress             overlay             swarm               
87b55179074b        none                null                local               
4t1qorqwy0b7        proxy               overlay             swarm               
8gw6au90pift        soss                overlay             swarm               
[root@dev-swarm-1 ~]# 
```

`docker service inspect mmane-sl`

```
 "VirtualIPs": [
                {
                    "NetworkID": "244ob6bfjdwh5gkihkn17i7o1",
                    "Addr": "10.255.0.78/16"
                },
                {
                    "NetworkID": "8gw6au90pifte2feb4uhdj2jz",
                    "Addr": "10.0.0.68/24"
                },
                {
                    "NetworkID": "4t1qorqwy0b7fkzpkmdsz514c",
                    "Addr": "10.0.1.75/24"
                }

```

```
[root@dev-swarm-1 ~]# docker service ps mmane-sl
ID                         NAME        IMAGE                                                                NODE                           DESIRED STATE  CURRENT STATE           ERROR
de1pt0kcfbj8xvxn5lm8xkuyj  mmane-sl.1  servicelayer:developer  dev-swarm-5.domain.com  Running        Running 10 minutes ago  
[root@dev-swarm-1 ~]# 

```

So why would it through that error message?  Also I've checked the haproxy configuration and I can see that virtualhosts are added 
```

   acl url_mmane-sl80 path_beg /
    acl domain_mmane-sl hdr_end(host) -i mmane-sl.domain.com
    use_backend mmane-sl-be80 if url_mmane-sl80 domain_mmane-sl


backend mmane-sl-be80
    mode http
    server mmane-sl mmane-sl:80
```

However when accessing via HTTP, I get 503 error

**Comment by vfarcic** [comment]
```
2017/01/20 12:54:55 Processing request /v1/docker-flow-proxy/reconfigure?distribute=true&port=80&serviceDomain=%2Ammane-sl.domain.com&serviceName=mmane-sl&servicePath=%2F

2017/01/20 12:54:55 Sending distribution request to http://10.0.1.78:8080/v1/docker-flow-proxy/reconfigure?distribute=false&port=80&serviceDomain=%2Ammane-sl.domain.com&serviceName=mmane-sl&servicePath=%2F

2017/01/20 12:54:55 Processing request /v1/docker-flow-proxy/reconfigure?distribute=false&port=80&serviceDomain=%2Ammane-sl.domain.com&serviceName=mmane-sl&servicePath=%2F
```

Those are not error messages but logging of the requests received or sent by the proxy. The first one means that it received a request to reconfigure the service `mmane-sl`. Since the request contains `distribute=true`, it means that it should not process it but send to all proxy replicas. The second logs entry means that it is resending the request but, this time, without distribute=true. Finally, the last entry is the final reconfigure request that ends up with the change of the configuration and reload of the proxy.

None of those entries is an error output. Does the proxy work after those requests? Can you send a request to the proxy and is it forwarding it to the `mmane-sl` service?

**Comment by getvivekv** [comment]
I apologize, looks like I didn't copy the error part. 
But the error message was 

"Could not reach the service. Is the service running and connected to the same network as the proxy"

**Comment by vfarcic** [comment]
That's not necessarily and error. When a Swarm service is created, it is only an entry in the scheduler. Some time passes from the moment a service is created and until replicas are running (Docker needs to pull the image, run health checks, and so on). If it's a small image, it can take a few milliseconds, while for big ones it can take seconds, or minutes.

In it's simplified form, the flow is as follows.

1. The Listener detects a new service.
2. The Listener sends a request to the Proxy to reconfigure itself with the information about the new service.
3. The Proxy checks whether the service is working. If it isn't, it does not reconfigure itself and sends an error message to whoever sent a reconfigure request.
4. If the Listener gets an error response from the Proxy, it retries the request from the step 2 unless a timeout is reached. In case of a timeout, the loop ends and the Proxy will not be reconfigured.

In other words, the service might never work and you should see quite a few messages like that one. More likelly, the service was only temporarily not working until Swarm pulled the images, run the containers, healthchecks, and so on.

Was your service reconfigured after that message? If the Proxy started serving requests, everything worked correctly. If that's the case, we should probably change the message you reported. Maybe something like:

"Could not reach the service. Is the service running and connected to the same network as the proxy? Please confirm that the service current state is running."

I would greatly appreciate if you have any suggestion how to formulate the message in a way that it is not misleading.

**Comment by getvivekv** [comment]
Is it possible to disable the (3) point - to NOT check whether the service is working and simply reconfigure itself? What I found that the service is actually working, I can ssh into the container and `docker service ls` and `docker service ps service_name` shows everything is up and running.

**Comment by vfarcic** [comment]
If the service was not running at all, you would see a long list of those messages since it retries the same request over and over again until the service starts working or it reaches timeout. The main question is not whether the service works in general but whether it worked at the time the listener sent a request to the proxy. If it was, there is a bug or networking is not configured correctly.

Can you please let me know how many of those messages you received for a single service and whether the proxy was not reconfigured at all?

During the initial stages of the project, proxy was not checking whether the destination service is running or not. It would change HAProxy configuration without any validation. A few users requested the validation so I added it. It was a good decision since it allowed them to start redirecting traffic to
the service only after it becomes available.

Please note that if you disable that validation, the listener will not retry requests since it will not know whether there is something wrong with the service or not.

I just added the `SKIP_ADDRESS_VALIDATION` env. variable. Set it to `true` and validation will be disabled.

All that being said, I feel we are moving away from the main topic of this issue.

 acl url_mmane-sl80 path_beg /
    acl domain_mmane-sl hdr_end(host) -i mmane-sl.domain.com
    use_backend mmane-sl-be80 if url_mmane-sl80 domain_mmane-sl


backend mmane-sl-be80
    mode http
    server mmane-sl mmane-sl:80
> However when accessing via HTTP, I get 503 error

**Comment by vfarcic** [comment]
Closing due to inactivity.
